### PR TITLE
Add a meaningful pull request template

### PR DESCRIPTION
**Pull request by patkar**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Tests pass? | n/a
| Fixed tickets | none
| License | MIT
| Doc PR | sulu-io/sulu-docs#180

#### TODO

- [x] Update the contributing docs

#### Description

This PR adds a meaningful pull request template, which defines a structure to encourage high quality PRs, which are understandable for new collaborators at every time. This supersedes and closes #2021 too.

The license is required to make sure all contributions are licensed under MIT. Otherwise it is a little bit unsure, which license a contribution has.
Another solution would be to require signing contribution license agreements, but IMHO the license line in the PR should be preferred.

#### Purpose

Most of the PRs in this repo have less or no description at all. Thus it is very hard to follow the changes as an outside collaborator. And as @webmozart is already questioning most of the issues and pull requests to add a meaningful description, this is the next logical step.

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+| Q | A
+| --- | ---
+| Bug fix? | [yes|no]
+| New feature? | [yes|no]
+| BC breaks? | [yes|no]
+| Deprecations? | [yes|no]
+| Tests pass? | [yes|no]
+| Fixed tickets | [comma separated list of tickets fixed by the PR -> `fixes #num`]
+| License | MIT
```
Comment (patkar) [inline .github/PULL_REQUEST_TEMPLATE.md]: The license is required to make sure all contributions are licensed under MIT. Otherwise it is a little bit unsure, which license a contribution has.

Another solution would be to require signing contribution license agreements, but IMHO the license line in the PR should be preferred.

**Comment by wachterjohannes** [comment]
@patkar i have already started a pull-request template https://github.com/sulu-io/sulu/pull/2021 maybe we can "combine" them

**Comment by wachterjohannes** [comment]
in my opinion this template is very good. @webmozart @chirimoya @danrot what do you think?

**Comment by danrot** [comment]
@patkar Is the license field really necessary? Sulu is licensed as MIT, so if somebody contributes code it's automatically MIT too (I am not a law expert, but this would make sense to me).

I would also omit Tests pass, that's what the PR checks are for. Whatever is written in there doesn't mean anything because of that.

And I wouldn't put the options (`[yes|no]`) into the template. Same goes for the text under description and purpose, because it is very annoying to delete them for every PR being created.

**Comment by wachterjohannes** [comment]
@danrot but i think you should think about what writing and we see that it hasn´t worked very good until now. we should give it a try and the template is also not only for us which know what todo when we create a PR its also like @patkar says for third-party developer which maybe contribute only once and don´t want to read an annoying doc to know what they should write into a PR. so +1 for remove license but -1 for remove example data.

**Comment by patkar** [comment]
The license field should make sure, contributed code is MIT. Without it, this is no clear and it doesn't matter if the project is MIT. Symfony does the same.

The tests pass field in the checklist is more dedicated to a yes/no for the new tests created and covered in the code of the PR. If no, it triggers a TODO to get the tests or rather the code right. See the documentation PR.

The table is in use for Symfony day by day.

The example/help text is a help for first timers, which encourages quality pull requests, but also for daily committers. Thereby you have to rethink to submit good descriptions and co while deleting the help text every time :neckbeard:

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
+| Q | A
+| --- | ---
+| Bug fix? | [yes|no]
+| New feature? | [yes|no]
+| BC breaks? | [yes|no]
+| Deprecations? | [yes|no]
+| Tests pass? | [yes|no]
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd remove this line. We use CI systems which automatically verify a PR's integrity. This flag here is most likely out of date with the actual contents of the PR.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+| Fixed tickets | [comma separated list of tickets fixed by the PR -> `fixes #num`]
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd make this simpler and more obvious:

`[fixes #abc, fixes #xyz]`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+| Related PRs | [comma separated list of related PRs, if needed]
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: again:

`[#345, #789]`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+| License | MIT
+| Documentation PR | [The reference to the documentation PR if any]
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: again:

`[Example: sulu/sulu-docs#123 - leave empty if not needed]`

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+#### TODO
+
+If you can't met the requirements of the table below, create a short todo
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: "meet", ["to-do list"](https://en.wikipedia.org/wiki/Wikipedia:To-do_list)

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+| New feature? | [yes|no]
+| BC breaks? | [yes|no]
+| Deprecations? | [yes|no]
+| Tests pass? | [yes|no]
+| Fixed tickets | [comma separated list of tickets fixed by the PR -> `fixes #num`]
+| Related PRs | [comma separated list of related PRs, if needed]
+| License | MIT
+| Documentation PR | [The reference to the documentation PR if any]
+
+#### TODO
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: To Do

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+If you can't met the requirements of the table below, create a short todo
+list and add the relevant parts:
+
+- [ ] fix the tests as they have not been updated yet
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd uppercase the first characters, as you did in your own PR.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+because you want early feedback on your work, add an item to the todo list:
+
+- [ ] finish the code
+- [ ] gather feedback for my changes
+
+As long as you have items in the todo list, please prefix the pull request
+title with `[WIP]`.
+Remove the TODO section if your pull request is complete and has no tasks.
+
+#### Description
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: See discussion in https://github.com/sulu-io/sulu-docs/pull/181

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (alexander-schranz) [inline .github/PULL_REQUEST_TEMPLATE.md]: Maybe change to `Added Tests?`

Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: Again, this won't be updated so it's retrospectively useless. I suspect we simply need to remind them to add tests.

A GitHub bot that automatically checks this would be awesome...

Comment (patkar) [inline .github/PULL_REQUEST_TEMPLATE.md]: As said: The tests pass field in the checklist is more dedicated to a yes/no for the new tests created and covered in the code of the PR. If no, it triggers a TODO to get the tests or rather the code right. See the documentation PR.

In other cases I updated it to shortly describe why travis fails on other stuff, but my additions pass ^^

**Comment by patkar** [comment]
Fixed the typos and checklist examples :)

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (danrot) [inline .github/PULL_REQUEST_TEMPLATE.md]: The problem is that it doesn't "trigger" anything, but only in a hypothetical process. People won't look at that field, and whatever is written here can't be trust.

Comment (patkar) [inline .github/PULL_REQUEST_TEMPLATE.md]: How about `Test coverage?`, so you know on the first sight if the PR has tests or not at least?

Comment (danrot) [inline .github/PULL_REQUEST_TEMPLATE.md]: Test coverage could also be given in case you are only doing a refactoring, and the tests have already been written before. For me it doesn't make any sense to add this line, whatever is written in here can't be taken for granted...

Comment (patkar) [inline .github/PULL_REQUEST_TEMPLATE.md]: @webmozart @alexander-schranz what do you say?

Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: I still think this line should be removed. Just look at the Symfony repository, there are plenty of PRs where this says "yes" and Travis fails. People don't think about it, so I doubt that they will question whether their PR contains tests and add them if it doesn't.

One solution would be to add "PLEASE ADD TESTS!" to the contents of this template. The problem is that people don't read the template anymore once they have seen it two or three times. It rather becomes a habit of removing pre-filled text and adding your own.

IMO what we need is someone or something that reminds them of adding the tests outside of this template. This could be as simple as commenting "Could you please add tests?" or maybe something more sophisticated, like a bot for this sort of check.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+| New feature? | [yes|no]
+| BC breaks? | [yes|no]
+| Deprecations? | [yes|no]
+| Tests pass? | [yes|no]
+| Fixed tickets | [Example: fixes #123, fixes #456]
+| Related PRs | [Example: #345, #789]
+| License | MIT
+| Documentation PR | [Example: sulu-io/sulu-docs#123 - leave empty if not needed]
+
+#### TODO
```
Comment (webmozart) [inline .github/PULL_REQUEST_TEMPLATE.md]: I would rename this to "To Do" and move this section to the very end. It's the least important section of all.

**Comment by webmozart** [comment]
The license is actually really important. Just because the project is MIT, that doesn't automatically make each contribution MIT AFAIK.

I added an alternative proposal in #2085.

**Comment by danrot** [comment]
Ok, when thinking about it, the license is ok, it also doesn't need to be touched anyway. And I like @webmozart approach better, because it is a lot shorter :-)

**Comment by patkar** [comment]
Actually removed the tests line, for the other points it's now a voting between those two templates :neckbeard:

**Comment by webmozart** [comment]
ping @wachterjohannes @dantleech @chirimoya

**Comment by wachterjohannes** [comment]
we should go with one i like the approach to give instructions in the template.

**Comment by chirimoya** [comment]
@patkar thanks for your effort, but we will merge https://github.com/sulu-io/sulu/pull/2085. We prefer the simplicity and concise descriptions.
